src/fe-common/core/fe-common-core.c

```diff
@@ -461,26 +462,63 @@ void fe_common_core_finish_init(void)
 
 gboolean strarray_find_dest(char **array, const TEXT_DEST_REC *dest)
 {
+	const WindowType type = window_item_get_type(dest->window->active);
+	GSList *targets = NULL, *iterator = NULL;
+	gboolean found = FALSE;
+
 	g_return_val_if_fail(array != NULL, FALSE);
 
+	// we ignore all targets
 	if (strarray_find(array, "*") != -1)
 		return TRUE;
-
-	if (strarray_find(array, dest->target) != -1)
+	// exit if not a channel or query
+	else if (type & WI_TYPE_OTHER)
+		return FALSE;
+	// we ignore all channels
+	else if (type & WI_TYPE_CHANNEL && strarray_find(array, "#") != -1)
+		return TRUE;
+	// Is this a dcc chat?
+	else if (type & WI_TYPE_DCCCHAT && strarray_find(array, "=") != -1)
+		return TRUE;
+	// Is this a private query window?
+	else if (type & WI_TYPE_PRIVMSG && strarray_find(array, "@") != -1)
 		return TRUE;
 
-	if (dest->server_tag != NULL) {
-		char *tagtarget = g_strdup_printf("%s/%s", dest->server_tag, "*");
-		int ret = strarray_find(array, tagtarget);
-		g_free(tagtarget);
-		if (ret != -1)
-			return TRUE;
+	g_return_val_if_fail(dest->server_tag != NULL, FALSE);
+
+	char *prefix = g_strdup_printf("%s/", dest->server_tag);
+	if (strarray_find_prefix(array, prefix) == -1) {
+		g_free(prefix);
+		return FALSE;
+	}
+
+	g_free(prefix);
 
-		tagtarget = g_strdup_printf("%s/%s", dest->server_tag, dest->target);
-		ret = strarray_find(array, tagtarget);
+	// create a list of types to look for
+	targets = g_slist_append(targets, g_strdup("*"));
```

Comment:
There's no need to strdup anything since the string literals have a ∞ lifetime and `dest->target` is definitely alive when it's feed into printf never to be used again.